scripts/dts/edtlib.py

```diff
@@ -0,0 +1,1626 @@
+        # Creates self._compat2binding. This is a dictionary that maps
+        # (<compatible>, <bus>) tuples (both strings) to (<binding>, <path>)
+        # tuples. <binding> is the binding in parsed PyYAML format, and <path>
+        # the path to the binding (nice for binding-related error messages).
+        #
+        # For example, self._compat2binding["company,dev", "can"] contains the
+        # binding/path for the 'company,dev' device, when it appears on the CAN
+        # bus.
+        #
+        # For bindings that don't specify a bus, <bus> is None, so that e.g.
+        # self._compat2binding["company,notonbus", None] contains the binding.
+        #
+        # Only bindings for 'compatible' strings that appear in the device tree
+        # are loaded.
+
+        dt_compats = _dt_compats(self._dt)
+        self._binding_paths = _binding_paths(bindings_dir)
+
+        # Add '!include foo.yaml' handling.
+        #
+        # Do yaml.Loader.add_constructor() instead of yaml.add_constructor() to be
+        # compatible with both version 3.13 and version 5.1 of PyYAML.
+        #
+        # TODO: Is there some 3.13/5.1-compatible way to only do this once, even
+        # if multiple EDT objects are created?
+        yaml.Loader.add_constructor("!include", self._binding_include)
+
+        self._compat2binding = {}
+        for binding_path in self._binding_paths:
+            compat = _binding_compat(binding_path)
+            if compat in dt_compats:
+                binding = _load_binding(binding_path)
+                self._compat2binding[compat, _binding_bus(binding)] = \
+                    (binding, binding_path)
+
+    def _binding_include(self, loader, node):
+        # Implements !include. Returns a list with the YAML structures for the
+        # included files (a single-element list if the !include is for a single
+        # file).
+
+        if isinstance(node, yaml.ScalarNode):
+            # !include foo.yaml
+            return [self._binding_include_file(loader.construct_scalar(node))]
+
+        if isinstance(node, yaml.SequenceNode):
+            # !include [foo.yaml, bar.yaml]
+            return [self._binding_include_file(filename)
+                    for filename in loader.construct_sequence(node)]
+
+        _yaml_inc_error("unrecognised node type in !include statement")
+
+    def _binding_include_file(self, filename):
+        # _binding_include() helper for loading an !include'd file. !include
+        # takes just the basename of the file, so we need to make sure there
+        # aren't multiple candidates.
+
+        paths = [path for path in self._binding_paths
+                 if os.path.basename(path) == filename]
+
+        if not paths:
+            _yaml_inc_error("'{}' not found".format(filename))
+
+        if len(paths) > 1:
+            _yaml_inc_error("multiple candidates for '{}' in !include: {}"
+                            .format(filename, ", ".join(paths)))
```

Comment:
This prevents binding overload by application. An application (OOT) driver may need a special binding.
Using the first file in the list and reporting a warning would help in this case. An application can arrange to have it's bindings directory first in the list of directories searched.

Reply:
The old code also gives an error here, so it's not a behavior change at least.

Reply:
Yes it is like in the old code. Just a remark. I patched it in my project to have an easy way to develop the OOT drivers.